apps/emqx_node_rebalance/src/emqx_node_rebalance.erl

```diff
@@ -0,0 +1,411 @@
+    {ok, State, Data}.
+
+%%--------------------------------------------------------------------
+%% internal funs
+%%--------------------------------------------------------------------
+
+enable_rebalance(#{opts := Opts} = Data) ->
+    Nodes = maps:get(nodes, Opts),
+    ConnCounts = multicall(Nodes, {?MODULE, connection_count, []}),
+    SessCounts = multicall(Nodes, {?MODULE, session_count, []}),
+    {_, Counts} = lists:unzip(ConnCounts),
+    Avg = avg(Counts),
+    {DonorCounts,
+     RecipientCounts} = lists:partition(
+                          fun({_Node, Count}) ->
+                                  Count >= Avg
+                          end,
+                          ConnCounts),
+    ?LOG(warning, "Enabling rebalance: ConnCounts=~p, DonorCounts=~p, RecipientCounts=~p",
+         [ConnCounts, DonorCounts, RecipientCounts]),
+    {DonorNodes, _} = lists:unzip(DonorCounts),
+    {RecipientNodes, _} = lists:unzip(RecipientCounts),
+    case need_rebalance(DonorNodes, RecipientNodes, ConnCounts, SessCounts, Opts) of
+        false -> {error, nothing_to_balance};
+        true ->
+            _ = multicall(DonorNodes, {emqx_node_rebalance_agent, enable, [self()]}),
+            {ok, Data#{donors => DonorNodes,
+                       recipients => RecipientNodes,
+                       initial_conn_counts => maps:from_list(ConnCounts),
+                       initial_sess_counts => maps:from_list(SessCounts)}}
+    end.
+
+disable_rebalance(#{donors := DonorNodes}) ->
+    _ = multicall(DonorNodes, {emqx_node_rebalance_agent, disable, [self()]}),
+    ok.
+
+evict_conns(#{donors := DonorNodes, recipients := RecipientNodes, opts := Opts} = Data) ->
+    DonorNodeCounts = multicall(DonorNodes, {?MODULE, connection_count, []}),
+    {_, DonorCounts} = lists:unzip(DonorNodeCounts),
+    RecipientNodeCounts = multicall(RecipientNodes, {?MODULE, connection_count, []}),
+    {_, RecipientCounts} = lists:unzip(RecipientNodeCounts),
+
+    DonorAvg = avg(DonorCounts),
+    RecipientAvg = avg(RecipientCounts),
+    Thresholds = thresholds(conn, Opts),
+    NewData = Data#{donor_conn_avg => DonorAvg,
+                    recipient_conn_avg => RecipientAvg,
+                    donor_conn_counts => maps:from_list(DonorNodeCounts),
+                    recipient_conn_counts => maps:from_list(RecipientNodeCounts)},
+    case within_thresholds(DonorAvg, RecipientAvg, Thresholds) of
+        true -> ok;
+        false ->
+            ConnEvictRate = maps:get(conn_evict_rate, Opts),
+            NodesToEvict = nodes_to_evict(RecipientAvg, DonorNodeCounts),
+            ?LOG(warning, "Node rebalance, evict_conns, nodes=~p, counts=~p",
+                 [NodesToEvict, ConnEvictRate]),
+            _ = multicall(NodesToEvict, {emqx_eviction_agent, evict_connections, [ConnEvictRate]}),
+            {continue, NewData}
+    end.
+
+evict_sessions(#{donors := DonorNodes, recipients := RecipientNodes, opts := Opts} = Data) ->
```

Comment:
For now we have nowhere to extend these two cases and I desided to leave them as is
* to avoid increasing ux complexity after adding hierarchy into parameters (`abs_sess_*`, `abs_conn_*`)
* following the rule of three strikes 😅 (while we have two)